File: src/cleansia_ios/CleansiaPartner/Sources/Features/Profile/Bank/BankSectionViewModel.swift
import CleansiaCore
import CleansiaPartnerApi
import Combine
import Foundation

/// The payout destination is captured as the parts a Czech or Slovak cleaner reads off their
/// statement — prefix, account number, bank code — because the server derives the IBAN from
/// them. Everything the server can check (the account checksum, the bank code, whether a
/// supplied IBAN agrees, whether a card number was typed in) is left to the server: it owns
/// the rules, its answers are wired to `error.validation.payout.*`, and a second copy of a
/// checksum here would gate a cleaner's income on a client that disagrees.
struct BankForm: Equatable {
    var employeeId = ""
    var bankCountryId: String?
    var accountPrefix = "" {
        didSet { accountPrefix = accountPrefix.payoutDigits }
    }

    var accountNumber = "" {
        didSet { accountNumber = accountNumber.payoutDigits }
    }

    var bankCode = "" {
        didSet { bankCode = bankCode.payoutDigits }
    }

    var iban = "" {
        didSet { iban = iban.bankReference }
    }

    var swift = "" {
        didSet { swift = swift.bankReference }
    }

    var bankName = ""
    var holderName = ""

    /// The bank's country plus something that identifies the account — the rest is the server's call.
    var canSubmit: Bool {
        guard let bankCountryId, !bankCountryId.isBlank else { return false }
        return !accountNumber.isBlank || !iban.isBlank
    }
}

@MainActor
final class BankSectionViewModel: ViewModel {
    @Published private(set) var state: UiState<BankForm> = .loading
    @Published private(set) var action: ActionState = .idle
    @Published private(set) var countryOptions: [CleansiaDropdownOption] = []

    let saved = PassthroughSubject<Void, Never>()

    private let client: PartnerProfileClient
    private let snackbar: SnackbarController

    init(client: PartnerProfileClient, snackbar: SnackbarController) {
        self.client = client
        self.snackbar = snackbar
    }

    func load() async {
        state = .loading
        countryOptions = await (client.getAllCountries()).valueOrNil?.compactMap(Self.option) ?? []

        let employee: EmployeeItem
        switch await client.getCurrentEmployee() {
        case let .success(value):
            employee = value
        case let .failure(error):
            return failLoad(error)
        }

        let payout: MyPayoutDetails?
        switch await client.getMyPayoutDetails() {
        case let .success(value):
            payout = value
        case let .failure(error):
            return failLoad(error)
        }

        state = .loaded(BankForm(
            employeeId: employee.id ?? "",
            // The bank is usually in the country the cleaner lives in, so pre-fill it and
            // let them change it — the same zero-tap default the identification section uses.
            bankCountryId: payout?.bankCountryId ?? employee.countryId,
            accountPrefix: PayoutAccountSummary.withoutPadding(payout?.accountPrefix),
            accountNumber: PayoutAccountSummary.withoutPadding(payout?.accountNumber),
            bankCode: payout?.bankCode ?? "",
            iban: payout?.iban ?? "",
            swift: payout?.swift ?? "",
            bankName: payout?.bankName ?? "",
            holderName: payout?.holderName ?? ""
        ))
    }

    func update<Value>(_ keyPath: WritableKeyPath<BankForm, Value>, to value: Value) {
        guard case var .loaded(form) = state else { return }
        form[keyPath: keyPath] = value
        state = .loaded(form)
    }

    func save() async {
        guard case let .loaded(form) = state, !action.isSubmitting else { return }
        guard !form.employeeId.isBlank else {
            snackbar.showError(L10n.Profile.errorProfileNotLoaded)
            return
        }
        guard form.canSubmit else { return }

        action = .submitting
        let command = UpdateBankDetailsCommand(
            employeeId: form.employeeId,
            iban: form.iban.trimmedOrNil,
            bankCountryId: form.bankCountryId,
            accountPrefix: form.accountPrefix.trimmedOrNil,
            accountNumber: form.accountNumber.trimmedOrNil,
            bankCode: form.bankCode.trimmedOrNil,
            swift: form.swift.trimmedOrNil,
            bankName: form.bankName.trimmedOrNil,
            holderName: form.holderName.trimmedOrNil
        )
        switch await client.updateBankDetails(command) {
        case .success:
            action = .idle
            saved.send()
        case let .failure(error):
            action = .idle
            snackbar.showApiError(error)
        }
    }

    private func failLoad(_ error: ApiError) {
        snackbar.showApiError(error)
        state = .error(error)
    }

    private static func option(_ country: CountryListItem) -> CleansiaDropdownOption? {
        guard let id = country.id else { return nil }
        return CleansiaDropdownOption(id: id, label: country.localizedName())
    }
}

private extension String {
    var payoutDigits: String {
        filter(\.isNumber)
    }

    var bankReference: String {
        uppercased().filter { $0.isLetter || $0.isNumber }
    }
}

private extension ApiResult {
    var valueOrNil: Success? {
        try? get()
    }
}

File: src/cleansia_android/partner-app/src/main/java/cz/cleansia/partner/features/profile/BankSectionViewModel.kt
package cz.cleansia.partner.features.profile

import android.content.Context
import androidx.lifecycle.ViewModel
import androidx.lifecycle.viewModelScope
import cz.cleansia.core.network.ApiResult
import cz.cleansia.core.network.safeApiCall
import cz.cleansia.core.snackbar.SnackbarController
import cz.cleansia.core.ui.state.ActionState
import cz.cleansia.partner.R
import cz.cleansia.partner.api.client.CountryApi
import cz.cleansia.partner.api.model.CountryListItem
import cz.cleansia.partner.core.network.ApiErrorTranslator
import cz.cleansia.partner.data.profile.ProfileRepository
import dagger.hilt.android.lifecycle.HiltViewModel
import dagger.hilt.android.qualifiers.ApplicationContext
import kotlinx.coroutines.flow.MutableSharedFlow
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.SharedFlow
import kotlinx.coroutines.flow.StateFlow
import kotlinx.coroutines.flow.asSharedFlow
import kotlinx.coroutines.flow.asStateFlow
import kotlinx.coroutines.flow.update
import kotlinx.coroutines.launch
import kotlinx.serialization.json.Json
import javax.inject.Inject

/**
 * The payout destination is captured as the parts a Czech or Slovak cleaner reads off their
 * statement — prefix, account number, bank code — because the server derives the IBAN from
 * them. Everything the server can check (the account checksum, the bank code, whether a
 * supplied IBAN agrees, whether a card number was typed in) is left to the server: it owns
 * the rules, its answers are wired to `error_validation_payout_*`, and a second copy of a
 * checksum here would gate a cleaner's income on a client that disagrees.
 */
data class BankForm(
    val employeeId: String = "",
    val countries: List<CountryListItem> = emptyList(),
    val bankCountryId: String? = null,
    val accountPrefix: String = "",
    val accountNumber: String = "",
    val bankCode: String = "",
    val iban: String = "",
    val swift: String = "",
    val bankName: String = "",
    val holderName: String = "",
) {
    /** The bank's country plus something that identifies the account — the rest is the server's call. */
    val canSubmit: Boolean
        get() = !bankCountryId.isNullOrBlank() && (accountNumber.isNotBlank() || iban.isNotBlank())
}

sealed interface BankSectionUiState {
    data object Loading : BankSectionUiState
    data object Error : BankSectionUiState
    data class Loaded(val form: BankForm) : BankSectionUiState
}

@HiltViewModel
class BankSectionViewModel @Inject constructor(
    private val profileRepository: ProfileRepository,
    private val countryApi: CountryApi,
    private val errorTranslator: ApiErrorTranslator,
    private val snackbar: SnackbarController,
    private val json: Json,
    @ApplicationContext private val appContext: Context,
) : ViewModel() {

    private val _uiState = MutableStateFlow<BankSectionUiState>(BankSectionUiState.Loading)
    val uiState: StateFlow<BankSectionUiState> = _uiState.asStateFlow()

    private val _saveState = MutableStateFlow<ActionState>(ActionState.Idle)
    val saveState: StateFlow<ActionState> = _saveState.asStateFlow()

    private val _saved = MutableSharedFlow<Unit>(extraBufferCapacity = 1)
    val saved: SharedFlow<Unit> = _saved.asSharedFlow()

    init { load() }

    fun retry() = load()

    private fun load() {
        viewModelScope.launch {
            _uiState.value = BankSectionUiState.Loading
            val countries = (safeApiCall(json) { countryApi.countryGetOverview() } as? ApiResult.Success)
                ?.data.orEmpty()

            val employee = when (val result = profileRepository.getCurrentEmployee()) {
                is ApiResult.Success -> result.data
                is ApiResult.Error -> return@launch failLoad(errorTranslator.translate(result.error))
            }
            val payout = when (val result = profileRepository.getPayoutDetails()) {
                is ApiResult.Success -> result.data
                is ApiResult.Error -> return@launch failLoad(errorTranslator.translate(result.error))
            }

            _uiState.value = BankSectionUiState.Loaded(
                BankForm(
                    employeeId = employee.id.orEmpty(),
                    countries = countries,
                    // The bank is usually in the country the cleaner lives in, so pre-fill it and
                    // let them change it — the same zero-tap default the identification section uses.
                    bankCountryId = payout?.bankCountryId ?: employee.countryId,
                    accountPrefix = payout?.accountPrefix.withoutPayoutPadding(),
                    accountNumber = payout?.accountNumber.withoutPayoutPadding(),
                    bankCode = payout?.bankCode.orEmpty(),
                    iban = payout?.iban.orEmpty(),
                    swift = payout?.swift.orEmpty(),
                    bankName = payout?.bankName.orEmpty(),
                    holderName = payout?.holderName.orEmpty(),
                ),
            )
        }
    }

    private fun failLoad(message: String) {
        snackbar.showError(message)
        _uiState.value = BankSectionUiState.Error
    }

    fun onBankCountrySelected(id: String) = updateForm { it.copy(bankCountryId = id) }

    fun onAccountPrefixChange(v: String) = updateForm { it.copy(accountPrefix = v.digitsOnly()) }

    fun onAccountNumberChange(v: String) = updateForm { it.copy(accountNumber = v.digitsOnly()) }

    fun onBankCodeChange(v: String) = updateForm { it.copy(bankCode = v.digitsOnly()) }

    fun onIbanChange(v: String) = updateForm { it.copy(iban = v.asBankReference()) }

    fun onSwiftChange(v: String) = updateForm { it.copy(swift = v.asBankReference()) }

    fun onBankNameChange(v: String) = updateForm { it.copy(bankName = v) }

    fun onHolderNameChange(v: String) = updateForm { it.copy(holderName = v) }

    fun save() {
        val form = (_uiState.value as? BankSectionUiState.Loaded)?.form ?: return
        if (_saveState.value is ActionState.Submitting) return
        if (form.employeeId.isBlank()) {
            snackbar.showError(appContext.getString(R.string.error_profile_not_loaded))
            return
        }
        if (!form.canSubmit) return

        viewModelScope.launch {
            _saveState.value = ActionState.Submitting
            val result = profileRepository.updateBankDetails(
                employeeId = form.employeeId,
                bankCountryId = form.bankCountryId,
                accountPrefix = form.accountPrefix.trimmedOrNull(),
                accountNumber = form.accountNumber.trimmedOrNull(),
                bankCode = form.bankCode.trimmedOrNull(),
                iban = form.iban.trimmedOrNull(),
                swift = form.swift.trimmedOrNull(),
                bankName = form.bankName.trimmedOrNull(),
                holderName = form.holderName.trimmedOrNull(),
            )
            when (result) {
                is ApiResult.Success -> {
                    _saveState.value = ActionState.Idle
                    _saved.emit(Unit)
                }
                is ApiResult.Error -> {
                    _saveState.value = ActionState.Idle
                    snackbar.showError(errorTranslator.translate(result.error))
                }
            }
        }
    }

    private inline fun updateForm(transform: (BankForm) -> BankForm) {
        _uiState.update { state ->
            if (state is BankSectionUiState.Loaded) state.copy(form = transform(state.form)) else state
        }
    }
}

private fun String.digitsOnly(): String = filter { it.isDigit() }

private fun String.asBankReference(): String = uppercase().filter { it.isLetterOrDigit() }

private fun String.trimmedOrNull(): String? = trim().ifBlank { null }
